### PR TITLE
Use Query typedef from yarl for params

### LINKED ISSUE #8563: params type signature does not match the runtime
### Describe the bug

Yarl supports a much wider definition of query parameters, including sequences of strings. I have to ignore lines that pass those more sophisticated types to the aiohttp calls.

### To Reproduce

```
 hosts = ["a", "b"]
 async with ClientSession() as client:
        resp = await client.get(
            "https://api.example.com/",
            params={"hosts": hosts},
        )
```

This causes a type error, but works perfectly well at runtime.
```
$ python3
Python 3.11.8 (main, Mar  8 2024, 09:32:16) [GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import yarl
>>> yarl.URL( "https://api.example.com/").update_query({"hosts": ["a", "b"]})
URL('https://api.example.com/?hosts=a&hosts=b')
```

### Expected behavior

I expect that the type signature for params in aiohttp matches the type signature in yarl.update_query.

### Logs/tracebacks

```python-traceback
/home/lpetre/dev/provision.py
  /home/lpetre/dev/provision.py:65:42 - error: Argument of type "dict[str, list[str] | str]" cannot be assigned to parameter "params" of type "Mapping[str, str] | None" in function "get"
    Type "dict[str, list[str] | str]" is incompatible with type "Mapping[str, str] | None"
      "dict[str, list[str] | str]" is incompatible with "Mapping[str, str]"
        Type parameter "_VT_co@Mapping" is covariant, but "list[str] | str" is not a subtype of "str"
          Type "list[str] | str" is incompatible with type "str"
            "list[str]" is incompatible with "str"
      "dict[str, list[str] | str]" is incompatible with "None" (reportArgumentType)
```

### Python Version

```console
$ python --version
Python 3.11.8
```

### aiohttp Version

```console
$ python -m pip show aiohttp
Name: aiohttp
Version: 3.9.5
Summary: Async http client/server framework (asyncio)
Home-page: https://github.com/aio-libs/aiohttp
Author: 
Author-email: 
License: Apache 2
Location: /home/lpetre/dev/src/github.com/midjourney/image-generation/.venv/lib/python3.11/site-packages
Requires: aiosignal, attrs, frozenlist, multidict, yarl
Required-by: gcsfs
```

### multidict Version

```console
$ python -m pip show multidict
Name: multidict
Version: 6.0.5
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /home/lpetre/dev/src/github.com/midjourney/image-generation/.venv/lib/python3.11/site-packages
Requires: 
Required-by: aiohttp, yarl
```

### yarl Version

```console
$ python -m pip show yarl
Name: yarl
Version: 1.9.4
Summary: Yet another URL library
Home-page: https://github.com/aio-libs/yarl
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache-2.0
Location: /home/lpetre/dev/src/github.com/midjourney/image-generation/.venv/lib/python3.11/site-packages
Requires: idna, multidict
```

### OS

$ lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 11 (bullseye)
Release:        11
Codename:       bullseye

### Related component

Client

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow the aio-libs Code of Conduct

### PR DESCRIPTION
**Pull request by lpetre**
## What do these changes do?

The type signature for params does not match the functionality provided by yarl. Update the type signature to expose the full feature set of yarl. 

## Are there changes in behavior for the user?

No changes to the user, expect that type checking will now match the runtime functionality.

## Is it a substantial burden for the maintainers to support this?

The only burden I see is that the type signature in yarl is not exported in the package, which means these two files should be kept in sync. I would recommend yarl be updated to export the type signature and aiohttp could use that directly.

## Related issue number
Fixes #8563

## Checklist

- [x] I think the code is well written
- [ ] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [ ] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [ ] Add a new news fragment into the `CHANGES/` folder
  * name it `<issue_or_pr_num>.<type>.rst` (e.g. `588.bugfix.rst`)
  * if you don't have an issue number, change it to the pull request
    number after creating the PR
    * `.bugfix`: A bug fix for something the maintainers deemed an
      improper undesired behavior that got corrected to match
      pre-agreed expectations.
    * `.feature`: A new behavior, public APIs. That sort of stuff.
    * `.deprecation`: A declaration of future API removals and breaking
      changes in behavior.
    * `.breaking`: When something public is removed in a breaking way.
      Could be deprecated in an earlier release.
    * `.doc`: Notable updates to the documentation structure or build
      process.
    * `.packaging`: Notes for downstreams about unobvious side effects
      and tooling. Changes in the test invocation considerations and
      runtime assumptions.
    * `.contrib`: Stuff that affects the contributor experience. e.g.
      Running tests, building the docs, setting up the development
      environment.
    * `.misc`: Changes that are hard to assign to any of the above
      categories.
  * Make sure to use full sentences with correct case and punctuation,
    for example:
    ```rst
    Fixed issue with non-ascii contents in doctest text files
    -- by :user:`contributor-gh-handle`.
    ```

    Use the past tense or the present tense a non-imperative mood,
    referring to what's changed compared to the last released version
    of this project.

### CODE REVIEW

**Review by Dreamsorcerer — COMMENTED** [review]
That is overly complex. I suggest we merge this into 3.10, but then create another PR for yarl to export the type, and we use that in 3.11+.

**Comment by Dreamsorcerer** [comment]
bdraco exported Query in yarl. So, we probably want a new PR using that type (and increasing the minimum yarl version).

File: aiohttp/client.py [hunk 1]
```diff
@@ -24,7 +24,6 @@
     Generic,
     Iterable,
     List,
-    Mapping,
     Optional,
```
Comment (bdraco) [inline aiohttp/client.py]:
```suggestion
    Mapping,
    Optional,
```

File: aiohttp/typedefs.py [hunk 2]
```diff
@@ -8,13 +8,24 @@
...
 
+try:
+    # Available in yarl>=1.10.0
+    from yarl import Query, QueryVariable, SimpleQuery
+except ImportError:
+    SimpleQuery = Union[str, int, float]
+    QueryVariable = Union[SimpleQuery, "Sequence[SimpleQuery]"]
+    Query = Union[
+        None, str, "Mapping[str, QueryVariable]", "Sequence[Tuple[str, QueryVariable]]"
+    ]
```
Comment (bdraco) [inline aiohttp/typedefs.py]: This can be stripped back on in 3.11 / master once we bump yarl

File: aiohttp/typedefs.py [hunk 3]
```diff
@@ -8,13 +8,24 @@
...
     Tuple,
     Union,
 )
 
 from multidict import CIMultiDict, CIMultiDictProxy, MultiDict, MultiDictProxy, istr
 from yarl import URL
 
+try:
+    # Available in yarl>=1.10.0
+    from yarl import Query, QueryVariable, SimpleQuery
```
Comment (Dreamsorcerer) [inline aiohttp/typedefs.py]: Only Query is used?

```suggestion
    from yarl import Query
```

**Review by bdraco — APPROVED** [review]
Thanks @lpetre

**Pull request merged**